admin/test/test_release.py

```diff
@@ -661,3 +667,637 @@ def test_publish_non_release_fails(self):
+        """
+        Calling :func:`update_repo` uploads packages from a source repository
+        to S3.
+        """
+        aws = FakeAWS(
+            routing_rules={},
+            s3_buckets={
+                self.target_bucket: {},
+            },
+        )
+
+        repo_contents = {
+            'clusterhq-flocker-cli-0.3.3-0.dev.7.noarch.rpm': 'cli-package',
+            'clusterhq-flocker-node-0.3.3-0.dev.7.noarch.rpm': 'node-package',
+        }
+
+        self.update_repo(
+            aws=aws,
+            yum=FakeYum(),
+            rpm_directory=self.rpm_directory,
+            target_bucket=self.target_bucket,
+            target_key=self.target_key,
+            source_repo=self.create_fake_repository(files=repo_contents),
+            packages=self.packages,
+        )
+
+        self.assertDictContainsSubset(
+            {os.path.join(self.target_key, package): repo_contents[package]
+             for package in repo_contents},
+            aws.s3_buckets[self.target_bucket])
+
+    def test_metadata_uploaded(self):
+        """
+        Calling :func:`update_repo` uploads metadata for a source repository to
+        S3.
+        """
+        aws = FakeAWS(
+            routing_rules={},
+            s3_buckets={
+                self.target_bucket: {},
+            },
+        )
+
+        self.update_repo(
+            aws=aws,
+            yum=FakeYum(),
+            rpm_directory=self.rpm_directory,
+            target_bucket=self.target_bucket,
+            target_key=self.target_key,
+            source_repo=self.create_fake_repository(files={}),
+            packages=self.packages,
+        )
+
+        self.assertIn(
+            os.path.join(self.target_key, 'repodata', 'repomd.xml'),
+            aws.s3_buckets[self.target_bucket])
+
+    def test_repository_added_to(self):
+        """
+        Calling :func:`update_repo` does not delete packages which already
+        exist in S3.
+        """
+        existing_s3_keys = {
+            os.path.join(self.target_key, 'existing_package.rpm'): '',
+        }
+
+        aws = FakeAWS(
+            routing_rules={},
+            s3_buckets={
+                self.target_bucket: existing_s3_keys,
+            },
+        )
+
+        repo_contents = {
+            'clusterhq-flocker-cli-0.3.3-0.dev.7.noarch.rpm': 'cli-package',
+            'clusterhq-flocker-node-0.3.3-0.dev.7.noarch.rpm': 'node-package',
+        }
+
+        self.update_repo(
+            aws=aws,
+            yum=FakeYum(),
+            rpm_directory=self.rpm_directory,
+            target_bucket=self.target_bucket,
+            target_key=self.target_key,
+            source_repo=self.create_fake_repository(files=repo_contents),
+            packages=self.packages,
+        )
+
+        expected_keys = existing_s3_keys.copy()
+        expected_keys.update({
+            os.path.join(self.target_key, package): repo_contents[package]
+            for package in repo_contents})
```

Comment:
This seems superfluous; that is, the new packages being there are already covered by other tests.